### PR TITLE
feat: Introducing StreamingCredentialsProvider for token based authentication

### PR DESCRIPTION
**Pull request by ndyakov**
### Implementation Details and Key Changes
1. **New Authentication Package**
   - Introduced `auth` package to hold types related to authentication
   - Added `auth.Credentials` interface to facilitate integration with external credential sources
   - Implemented `basicAuth` for username/password credentials

2. **Streaming Credentials Provider**
   - Introduced `StreamingCredentialsProvider` interface for dynamic credential updates
   - Added support for managed identity scenarios where credentials are retrieved from external sources
   - Implemented `ReAuthCredentialsListener` for handling credential updates
   - Added comprehensive test coverage for all authentication components

3. **Documentation Updates**
   - Added detailed Authentication section to README.md
   - Documented all credential provider types and their priority order:
     - Streaming Credentials Provider (highest priority)
     - Context-based Credentials Provider
     - Regular Credentials Provider
     - Username/Password fields (lowest priority)
   - Added example of Entra ID (Azure AD) authentication integration

4. **Hooks System Enhancement**
   - Refactored `hooksMixin` to propagate hooks to child connections
   - Hooks are now triggered prior to the initialization process
   - This change addresses reported issues with hook propagation

5. **Testing**
   - Added unit tests for all authentication components
   - Tested credential provider priority order
   - Tested streaming credentials provider with mock implementation
   - Tested error handling and edge cases

## Security Considerations
**Important Note**: The hooksMixin refactor will now propagate hooks to child connections and trigger them prior to the initialization process. This change may have security implications, particularly regarding the visibility of authentication commands. For example, Redis MONITOR doesn't report AUTH commands, and this change could potentially expose sensitive authentication information through hooks. This should be carefully considered and potentially documented for users implementing custom hooks.

## Related Issues
- Fixes issues with hook propagation to child connections 
  - #3267 
  - #2680 
- Improves documentation for authentication system

### CODE REVIEW

File: options.go [hunk 1]
```diff
@@ -531,7 +582,6 @@ func newConnPool(
 		PoolFIFO:        opt.PoolFIFO,
 		PoolSize:        opt.PoolSize,
 		PoolTimeout:     opt.PoolTimeout,
-		DialTimeout:     opt.DialTimeout,
```
Comment (vladvildanov) [inline options.go:534]: Is it an internal one or part of a public API?

Comment (ndyakov) [reply options.go:534]: Actually, good catch, that was removed during the resolution of the merge conflicts. Should be put back

**Review by vladvildanov — DISMISSED** [review]

**Review by elena-kolevska — COMMENTED** [review]
LGTM but I think the PR title is misleading. From what I understand, this PR doesn't provide the possibility for authentication with a `StreamingCredentialsProvider` just yet, instead it introduces the necessary interfaces and already prepares the ground by refactoring the basic auth we were using to use the new Credentials interface.

**Comment by ndyakov** [comment]
Reverting this back to Draft, will continue working on it before `v9.8.0-beta.2`

File: auth/reauth_credentials_listener.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+// If the reAuth function returns an error, it calls the onErr function with the error.
+func (c *ReAuthCredentialsListener) OnNext(credentials Credentials) {
+	if c.reAuth != nil {
+		err := c.reAuth(credentials)
+		if err != nil {
+			if c.onErr != nil {
+				c.onErr(err)
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (ndyakov) [reply auth/reauth_credentials_listener.go:24]: I agree on calling `c.OnError`. As for early returns - i don't have a preference when the function is as short as this, but makes sense if  we are gonna use them elsewhere and would like to be consistent. Thank you for the suggestions!

File: redis.go [hunk 3]
```diff
@@ -282,36 +285,107 @@ func (c *baseClient) _getConn(ctx context.Context) (*pool.Conn, error) {
...
+	} else if c.opt.CredentialsProviderContext != nil {
+		username, password, err = c.opt.CredentialsProviderContext(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get credentials from context provider: %w", err)
+		}
+	} else if c.opt.CredentialsProvider != nil {
+		username, password = c.opt.CredentialsProvider()
+	} else if c.opt.Username != "" || c.opt.Password != "" {
+		username, password = c.opt.Username, c.opt.Password
+	}
```
[the remark this answers is not in the document]
Comment (ndyakov) [reply redis.go:383]: Although both versions are identical, I think the `if / else if/ else` makes it easier to follow the priority of the credentials providers. This is a personal preference, let me know if you have strong opinion on the topic.

**Review by elena-kolevska — APPROVED** [review]
One comment and small question, but in general lgtm

**Review by bobymicroby — DISMISSED** [review]
LGTM! Great job on this, the PR looks awesome!  I was initially wondering about omitting the `hooksMixin` when creating the connection for re-authentication (`cn := newConn(c.opt, connPool, nil)` in `reAuthConnection`). However, it makes sense as a defensive measure to prevent potential infinite loops if a hook itself indirectly triggered a re-authentication.
